File: Course_1-Algorithmic_Toolbox/Week-5/Excercise_Challenges/3_edit_distance/edit_distance.py
import numpy as np

def edit_distance(s1, s2):
    n, m = len(s1), len(s2)
    D = np.zeros((n + 1, m + 1))

    for i in range(m + 1):
        D[0][i] = i
    for i in range(n + 1):
        D[i][0] = i

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            match = D[i - 1][j - 1]
            mismatch = D[i - 1][j - 1] + 1
            insertion = D[i][j - 1] + 1
            deletion = D[i - 1][j] + 1

            if s1[i - 1] == s2[j - 1]:
                D[i][j] = min(insertion, deletion, match)
            else:
                D[i][j] = min(insertion, deletion, mismatch)

    return int(D[n][m])

if __name__ == "__main__":
    a = input()
    b = input()
    print(edit_distance(a, b))

File: Course_1-Algorithmic_Toolbox/Week-5/Excercise_Challenges/3_edit_distance/edit_distance.cpp
#include <iostream>
#include <string>
using namespace std;

int min3(int a, int b, int c)
{
    return min(min(a, b), c);
}

int edit_distance(const string &str1, const string &str2)
{
    int n = str1.length(), m = str2.length();
    int D[n + 1][m + 1];

    for(int i = 0; i <= m; ++i)
        D[0][i] = i;

    for(int i = 0; i <= n; ++i)
        D[i][0] = i;

    for(int i = 1; i <= n; ++i){
        for(int j = 1; j <= m; ++j){
            int match = D[i - 1][j - 1];
            int mismatch = D[i - 1][j - 1] + 1;
            int deletion = D[i - 1][j] + 1;
            int insertion = D[i][j - 1] + 1;

            if(str1[i - 1] == str2[j - 1])
                D[i][j] = min3(insertion, deletion, match);
            else
                D[i][j] = min3(insertion, deletion, mismatch);
        }
    }

    return D[n][m];
}

int main(void)
{
    string str1, str2;
    cin >> str1;
    cin >> str2;

    cout << edit_distance(str1, str2) << endl;
    return 0;
}
